modyn/evaluator/internal/pytorch_evaluator.py

```diff
@@ -142,38 +128,42 @@ def evaluate(self) -> None:
                 with torch.autocast(self._device_type, enabled=self._amp):
                     output = self._model.model(data)
 
-                    for metric in self._metrics:
+                    for metric in metrics:
                         if isinstance(metric, AbstractDecomposableMetric):
                             metric.evaluate_batch(target, output, batch_size)
 
-                    if self._contains_holistic_metric:
+                    if contains_holistic_metric:
                         y_true.append(target.detach().cpu())
                         y_score.append(output.detach().cpu())
 
-                self._num_samples += batch_size
+                num_samples += batch_size
 
         if len(y_true) > 0:
-            assert self._contains_holistic_metric  # We only track y_true in case of holistic metrics
+            assert contains_holistic_metric  # We only track y_true in case of holistic metrics
             y_true = torch.cat(y_true)
             y_score = torch.cat(y_score)
 
-            for metric in self._metrics:
+            for metric in metrics:
                 if isinstance(metric, AbstractHolisticMetric):
-                    metric.evaluate_dataset(y_true, y_score, self._num_samples)
+                    metric.evaluate_dataset(y_true, y_score, num_samples)
 
-        # We do this since we might also have just non-holistic metrics, in which case len(y_true) always is 0
-        for metric in self._metrics:
-            self._metric_result_queue.put((metric.get_name(), metric.get_evaluation_result()))
+        metric_result = []
+        for metric in metrics:
+            metric_result.append((metric.get_name(), metric.get_evaluation_result()))
+        self._metric_result_queue.put((interval_idx, metric_result))
```

Comment:
Just for my understanding: The interval idx stems from the index in the list of indices that the supervisor sends, correct? So the index is implicltly given by the supervisor, and the interval is not identified by the [start, end] timestamps, correct?

Reply:
yes. My thought is that [start, end] is somewhat complex, and also not unique in the semantics:
[0, 2000] and [None, 2000] means the same interval.

Also on the supervisor side, when collecting the evaluation results, it is easier to just index into the interval lists, rather than going through the list and matching the `[start, end]` pattern. What do you think?

Reply:
Maybe i am missing something but i dont see where at the supervisor we currently have this mapping. I guess my concern mostly is that we run into a discrepancy of indices between evaluator and supervisor. But ok if its very easy to have this mapping at the supervisor as well, then we can keep your approach

Reply:
no the indices between the evaluator and supervisor are consistent.
The supervisor naturally have a list of intervals, because it is the supervisor that put this interval list into the evaluation request and send the request.

How the indices are tracked on evaluator:
1. On `evaluator_grpc_servicer.py`, we enumerate over the original interval list and record the ids that didn't fail. https://github.com/eth-easl/modyn/blob/94efd22fb40dbdaf9c7adc5be2c63e3cb4f680c8/modyn/evaluator/internal/grpc/evaluator_grpc_servicer.py#L146
2. This `not_failed_interval_ids` is passed to the `EvaluationInfo` together with the original full interval list https://github.com/eth-easl/modyn/blob/94efd22fb40dbdaf9c7adc5be2c63e3cb4f680c8/modyn/evaluator/internal/grpc/evaluator_grpc_servicer.py#L197.
3. With these two fields, the `pytorch_evaluator` can only execute the not failed intervals and associate them with the right ids in the `metric_result_queue` https://github.com/eth-easl/modyn/blob/94efd22fb40dbdaf9c7adc5be2c63e3cb4f680c8/modyn/evaluator/internal/pytorch_evaluator.py#L157

Reply:
Can we let the supervisor encode the indices explicitly in the payload (not only by the position in an array)? Then the evaluator just refers to the id it was given from the payload if we build indices with enumeration no both sides that sounds like a potential pitfall where things start to fail silently, just because someone made wrong assumptions about interval ordering.